### PR TITLE
Custom window frame logic (support `ROWS`, `RANGE`, `PRECEDING` and `FOLLOWING` for window functions)

### PR DESCRIPTION
**Pull request by metesynnada**
****Which issue does this PR close?****

We offer a partial implementation of [windows with custom window frames](https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-WINDOW-FUNCTIONS) and improve the situation on [#361](https://github.com/apache/arrow-datafusion/issues/361).

As a team, this is our first contribution to Datafusion and we hope to contribute further to both Datafusion and Ballista in the future. Since we are creating a PR for this project for the first time, we would like to get feedback on how we are doing in terms of code quality, alignment with the project roadmap etc. We also would like to get your ideas on how to close this issue completely since we are providing a partial implementation as a first step. You can see which cases we cover in integration tests.

****Rationale for this change****

Datafusion currently does not support custom window frames, but it is on the [roadmap](https://arrow.apache.org/datafusion/user-guide/sql/sql_status.html).

****What changes are included in this PR?****

For now, we implemented `ROWS` and `RANGE` modes supporting `PRECEDING` and `FOLLOWING`.

As a draft, we currently do not support:

- `GROUPS` mode
- Timestamp ranges; e.g. `RANGE BETWEEN '1 day' PRECEDING AND '10 days' FOLLOWING` since the logical planner does not support the types other than an integer.
- Frame exclusion, i.e `EXCLUDE CURRENT ROW`

**Next steps**

- `GROUPS` mode implementation by extending `calculate_current_window` method.
- Frame exclusion, by logical planner extension and adapting  `calculate_current_window` method.

**Observations**

- Some aggregation function implementations are not generic, but use `f64`. This can create issues with statistical aggregation functions like `CORR(x, y)` when greater precision is required. Fortunately, they can be enhanced to support other data types similar to `SUM(x)` aggregation.
    
    Also, `evaluation()` of the `CovarianceAccumulator` should be
    
    ```
    @ -374,12 +374,6 @@ impl Accumulator for CovarianceAccumulator {
      };
    
      if count <= 1 {
    -      return Err(DataFusionError::Internal(
    -          "At least two values are needed to calculate covariance".to_string(),
    -      ));
    -  }
    -
    -  if self.count == 0 {
          Ok(ScalarValue::Float64(None))
      } else {
          Ok(ScalarValue::Float64(Some(self.algo_const / count as f64)))
    ```
    
    to become compatible with PostgreSQL. However, these issues are separate from this PR and we can discuss them under a new issue. For this reason, we deferred supporting functions like `CORR(x, y)` to the future.
    
- Since unstable sorting is used, some queries output different results than PostgreSQL. We use only unique columns for `ORDER BY` clauses while testing `ROWS` mode.

    An example query:

    ```sql
    SELECT c2, c3,
       SUM(c2) OVER(ROWS BETWEEN 3 PRECEDING AND 1 FOLLOWING) as summation2,
       SUM(c3) OVER(ORDER BY c2 ROWS BETWEEN 3 PRECEDING AND 1 FOLLOWING) as summation3,
       SUM(c3) OVER(ORDER BY c1 ROWS BETWEEN 3 PRECEDING AND 1 FOLLOWING) as summation4
    FROM test
    LIMIT 10;
    ```

    The Datafusion output is:

    ```sql
    +----+-----+------------+------------+------------+
    | c2 | c3  | summation2 | summation3 | summation4 |
    +----+-----+------------+------------+------------+
    | 1  | 12  | 2          | 132        | -13        |
    | 1  | 120 | 3          | 203        | 263        |
    | 1  | 71  | 4          | 118        | 447        |
    | 1  | -85 | 5          | 154        | 37         |
    | 1  | 36  | 5          | 43         | 358        |
    | 1  | -99 | 5          | 48         | -81        |
    | 1  | 125 | 5          | -31        | 247        |
    | 1  | -8  | 5          | 111        | 215        |
    | 1  | 57  | 5          | 3          | 238        |
    | 1  | -72 | 5          | 140        | 247        |
    +----+-----+------------+------------+------------+
    ```

    and in PostgreSQL as

    ```sql
    + --- + --- + ---------- + ---------- + ---------- +
    | c2  | c3  | summation2 | summation3 | summation4 |
    | --- | --- | ---------- | ---------- | ---------- |
    | 1   | -85 | 2          | -49        | -251       |
    | 1   | 36  | 3          | 71         | 330        |
    | 1   | 120 | 4          | 46         | 284        |
    | 1   | -25 | 5          | 149        | -184       |
    | 1   | 103 | 5          | 305        | -15        |
    | 1   | 71  | 5          | 323        | 251        |
    | 1   | 54  | 5          | 286        | 48         |
    | 1   | 83  | 5          | 255        | 166        |
    | 1   | -56 | 5          | 222        | -79        |
    | 1   | 70  | 5          | 180        | -233       |
    +-----+-----+------------+------------+------------+
    ```

- There is a minor problem in the logical planner, it should run

    ```sql
    SELECT 
      SUM(c2) OVER(ORDER BY c5, c6 RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) as sum
    FROM test;
    ```

    without a problem, however, it produces

    ```
    Error: Plan("With window frame of type RANGE, the order by expression must be of length 1, got 2")
    ```

### CODE REVIEW

**Comment by ozankabak** [comment]
@alamb, mentioning you since you showed interest in this. Once this gets some feedback, we will perform any fixes if necessary and then add `GROUPS` functionality so that this work can get merged.

Note that `EXCLUDE CURRENT ROW` and timestamp handling requires changes to planning logic, so we will leave those to another PR.

**Comment by alamb** [comment]
Thanks @ozankabak  -- I will try and find time to review this later this week, though it may not be until the weekend

**Comment by alamb** [comment]
I took a brief look through this PR, and it looks quite cool -- will provide detailed comments later. For now I have started the CI checks

**Comment by ozankabak** [comment]
Seems like we forgot the license header in one file 🙂 CI checks are passing except that. We will fix the license thing along with any other changes based on the review.

File: datafusion/core/tests/sql/window.rs
Comment (alamb) [inline datafusion/core/tests/sql/window.rs:506]: I am surprised this query has deterministic results -- I would expect the OVER clause needs an ORDER BY a well (otherwise the values that c4 will be computed could be different.

Something like:

```suggestion
               SUM(c4) OVER(ORDER BY c4 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING),\
```
(and the same for the OVER below as well)

File: datafusion/core/tests/sql/window.rs
Comment (alamb) [inline datafusion/core/tests/sql/window.rs:553]: One way to handle this type of non determinism would be add a where clause that filtered out any non unique values

File: datafusion/core/tests/sql/window.rs
Comment (alamb) [inline datafusion/core/tests/sql/window.rs:586]: Given that c9 is unique, all the window sizes here will be `1` so its coverage of the preceding and following is somewhat limited

```
❯ select count(c9), count(distinct c9) from aggregate_test_100;
+------------------------------+---------------------------------------+
| COUNT(aggregate_test_100.c9) | COUNT(DISTINCT aggregate_test_100.c9) |
+------------------------------+---------------------------------------+
| 100                          | 100                                   |
+------------------------------+---------------------------------------+
1 row in set. Query took 0.004 seconds.
```

File: datafusion/core/tests/sql/window.rs
Comment (alamb) [inline datafusion/core/tests/sql/window.rs:620]: I wonder if you ran the same queries in postgres? If not, no worries, but I am having a hard time validating these answers are correct (mostly because the `register_aggregate_csv` isn't a great dataset lol)

File: datafusion/expr/src/window_frame.rs
Comment (alamb) [inline datafusion/expr/src/window_frame.rs:95]: This code now always returns OK, right? it seems like the check on `bound` can be removed entirely, though perhaps I  am missing something. 


```rust
        } else {
            let units = value.units.into();
            Ok(Self {
                units,
                start_bound,
                end_bound,
            })
        }
```

File: datafusion/physical-expr/src/aggregate/correlation.rs
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/correlation.rs:156]: I think you can avoid these clones with something like

```rust
        self.stddev1.retract_batch(&values[0..1])?;
        self.stddev2.retract_batch(&values[1..2])?;
```

(aka slice the slice!)

File: datafusion/physical-expr/src/aggregate/covariance.rs
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/covariance.rs:288]: I haven't carefully reviewed this code

**Review by alamb — COMMENTED** [review]
@metesynnada  thank you very much for this contribution. The `retract_rows` is a very clever way to implement this feature. 

> As a team, this is our first contribution to Datafusion and we hope to contribute further to both Datafusion and Ballista in the future

That is wonderful news and welcome!

> We also would like to get your ideas on how to close this issue completely since we are providing a partial implementation as a first step. You can see which cases we cover in integration tests.

> As a draft, we currently do not support:

I think in general we have taken the approach of incrementally adding feature support. Thus if a feature is not entirely supported, implementing it incrementally (e.g. adding support for more of, but not all of the `OVER` clause) is both "allowed" and encouraged as it makes the PRs easier to review, given our limited reviewer bandwidth

While I am typing this, I would like to encourage you and your team to feel free to review / approve other PRs. One of the maintainers will review prior to merging and we have found that having community reviews helps with code quality, spreads knowledge, and makes the PR merge process more efficient.


T


cc @Jimexist  who contributed the initial implementation of windows functions

File: datafusion/physical-expr/src/window/aggregate.rs
Comment (alamb) [inline datafusion/physical-expr/src/window/aggregate.rs:350]: since we are iterating through the vec anyways above, rather than incrementally gathering `Vec<Result<_>>` you could use `?` and only push to `row_wise_results` on success. It might make this code slightly simpler

**Review by alamb — COMMENTED** [review]
I am still working through more of this code, but in summary I think it is very good, and well tested 👍  I went through the code in datafusion/physical-expr/src/window/aggregate.rs quite carefully and it looks really nice

If possible, it would be great to add some negative SQL level tests showing what types of SQL OVER clauses (e.g. with groups, and what types of bounds) are not supported (so for example we don't inadvertently plan them incorrectly)

My only real concern is about the growth of features of `ScalarValue` (and automatic coercion) but otherwise I think this PR is looking very close to mergable

For follow on work, in this project we typically file tickets to track the missing features and merge in the initial feature.  

So all in all, thank you @metesynnada  -- this is a great initial PR

File: datafusion/core/tests/sql/window.rs
Comment (jhorstmann) [reply datafusion/core/tests/sql/window.rs:506]: I believe the `ORDER BY c9` below makes this deterministic. Except when there are duplicates in `c9` with different values in `c4`, as you also note below. In that case it would probably be best to extend the `ORDER BY` with another column that makes it stable.

File: datafusion/common/src/bisect.rs
Comment (Jimexist) [inline datafusion/common/src/bisect.rs:31]: if this returns an index then it's a binary search instead of bisect?

**Review by Jimexist — COMMENTED** [review]
thanks for the contribution

i feel that bisect and scalar changes are better to be extracted and merge first

**Comment by ozankabak** [comment]
Thanks for the reviews, everybody! We are working on changes that address your reviews and plan to send them your way in a day or two.

File: datafusion/physical-expr/src/aggregate/sum.rs
Comment (metesynnada) [reply datafusion/physical-expr/src/aggregate/sum.rs:514]: Since we alter the summation operation and added the subtraction operation inside the `ScalarValue` implementation, we moved the relevant tests under the `scalar.rs`.

File: datafusion/physical-expr/src/window/aggregate.rs
Comment (metesynnada) [reply datafusion/physical-expr/src/window/aggregate.rs]: As per your suggestion, we have explicitly written all variants in the match. The variants returning error indicates the invalid window frames. These variants should be rejected in the Logical plan phase (they are indeed rejected). Hence those variants return DataFusionError::Internal. Thanks for the suggestion, listing the variants explicitly clarified the error cases.

**Comment by ozankabak** [comment]
Hello again! We resolved almost all the points you've mentioned. Additionally, we identified some challenging corner cases regarding NULL treatment and updated our code to handle these cases, along with new unit tests for these corner cases. Three main discussion points remain:

*PostgreSQL compatibility of new unit tests*

Expected values of all the new unit tests were sanity-checked against PostgreSQL. To remain in sync with PostgreSQL, we also think that it could be a good idea to add psql-parity tests for NULL-treatment cases. If you agree, let's make this subject of another PR. We can open an issue to track this (and will be happy to work on a PR to resolve it).

*Changes to `ScalarValue`*

At this point, we do not really have any new functionality or complexity added to this type. We just tidied up/moved some `ScalarValue`-only code that was scattered across multiple modules to `scalar.rs`. For example, the coercions you saw were already a part of the code here:

https://github.com/apache/arrow-datafusion/blob/87faf868f2276b84e63cad6721ca08bd79ed9cb8/datafusion/physical-expr/src/aggregate/sum.rs#L261

We agree that there should be no coercions in general at this stage -- any necessary work along these lines should be handled previously in the code path. However, we don't want to extend the scope of this PR to remove already-existing coercions from the codebase. If there is no issue about this, we are happy to open an issue for this and help resolve it in the near future along with similar issues like [#3511](https://github.com/apache/arrow-datafusion/issues/3511).

Some teaser: I have done some preliminary tests and it seems like removing all the coercions will not be a tough task at all. However, removing this kind of code always requires more careful testing than just preliminary testing and deserves a careful study/separate PR 🙂

*Bisect*

We considered using the [`partition_point`](https://doc.rust-lang.org/std/primitive.slice.html#method.partition_point)  function. This requires converting `&[ArrayRef]` to `&[&[ScalarValue]]`, where the inner `&[ScalarValue]` corresponds to each row (Because we are searching the insertion place of target `&[ScalarValue]` among rows). Unless we are missing something, doing this will require unnecessary memory usage and CPU time since it will involve copying. On the contrary, our implementation takes data in original columnar format and calculates rows from the index only when the comparison is done. Since we are doing *log(n)* comparisons on average, we need *log(n)* copies instead of *n* copies. If you have any suggestions for adding bisect functionality without changing the columnar orientation of the data, we can create an issue for this and help resolve it in a subsequent PR.

Thanks again for all the reviews. Excited to see this merge and get used!

**Comment by alamb** [comment]
Thanks @metesynnada  -- I'll check this out more carefully tomorrow or Friday

File: datafusion/core/tests/sql/window.rs
Comment (alamb) [inline datafusion/core/tests/sql/window.rs:662]: I think `sum` will produce the same value regardless of the window order -- perhaps using `first()` might add better coverage to these tests

File: datafusion/physical-expr/src/aggregate/sum.rs
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/sum.rs:329]: It took me a while to understand why this `count` field is necessary. I am a little worried that it might add overhead to the aggregate function (aka slow down aggregates). However, we can always implement a special case sum for window functions if this is a problem. 

I think adding some comments saying that the count is needed to distinguish the case of "no values" if all values are  removed via `remove_batches` would help future readers. 

I did try commenting it out and there was a test diff, so 👍  for the coverage.

**Review by alamb — APPROVED** [review]
Thanks again ! I went through this PR again, and I think it is ready to merge. 

> Expected values of all the new unit tests were sanity-checked against PostgreSQL. To remain in sync with PostgreSQL, we also think that it could be a good idea to add psql-parity tests for NULL-treatment cases. If you agree, let's make this subject of another PR. We can open an issue to track this (and will be happy to work on a PR to resolve it).

Yes, a follow on issue would be great. Thank you

> Currently, if an overflow happens, the whole process panics and quits as there is no machinery to handle this in the codebase. If you let us know of your thinking on this matter, we will be happy to help get this fixed in general in a subsequent PR so that other code can benefit too.

Likewise, I think follow on PRs are the right way to go -- thank you.

**Comment by alamb** [comment]
I'll plan to merge this tomorrow unless I hear otherwise or anyone else would like longer to review. Thanks again  @metesynnada and @ozankabak -- this looks epic

**Comment by ozankabak** [comment]
Thanks @alamb! In our first follow-up PR, we will tweak the test you noted to increase coverage and also enrich the comments to explain subtle points such as the one you found.

**Comment by alamb** [comment]
🚀  -- great first contribution

**Pull request merged**
